### PR TITLE
iptables: rename *tables-nft packages

### LINKED ISSUE #9540: firewall4: add dependency iptables-nft and ip6tables-nft?
version: reboot-19180-g73c6d8fd04
I have installed a fresh system. I have additional installed the mwan3 package.
I now have a problem with the firewall.

The mwa3 works with both firewall3(iptables) and firewall4(nft). For this to work, the rules for iptables must be created with iptables-legacy or for nft with iptables-nft.
The firewall4(nft) is installed by default and only iptables or ip6tables is in the dependency of mwan3.
This means that the package is satisfied when the dependency are resolved (PROVIDE:iptables or PROVIDE:ip6tables).

On my system, however, iptables-legacy(IPv4) and iptables-nft(IPv6) are now installed.
I would have expected that if the firewall4(nft) is installed then iptables-nfg(IPv4) and ip6tables(IPv6) is installed as a transitional solution.

As it is now, it is definitely bad. I think if we use firewall4(nft) we have to install iptables-nft and ip6tables-nft

### PR DESCRIPTION
**Pull request by feckert**
It only makes sense that we only have one PROVIDES definition for the
iptables packages, because a  mix of nft and legacy iptables backend
is not recommended at the same time.

Since the firewall4 is now the default firewall in openwrt, the
iptables-nft/ip6tables-nft backend for iptables must installed,
if a package wants to use the iptables/ip6tables commands.

@champtar @bluewavenet @hnyman 
In order to get further, I have now set up this pullreqest, as suggested in https://github.com/openwrt/openwrt/issues/9540, only set the PROVIDE for the iptables-nft and ip6tables-nft packages

Fixes #9540

### CODE REVIEW

**Comment by bluewavenet** [comment]
@feckert @hnyman @champtar @mans0n 
This also fixes:
https://github.com/openwrt/routing/issues/827

and makes redundant the PR:
https://github.com/openwrt/routing/pull/846

**Comment by champtar** [comment]
This will prevent everyone from easily switching back to fw3. Wouldn't it be enough to make sure nft variant is the preferred one ?

**Comment by feckert** [comment]
@bluewavenet 
> But do you actually need iptables-legacy to run fw3? Will it not work with iptables-nft? That is the point of iptables-nft, ie to support older packages that only understand the iptables api.

It is not recommended to mix iptables-nft and iptables-legacy commands.
If you must use old iptables commands you should use iptables-nft/ip6tables-nft for fw4 and iptables-legacy/ip6tables-legacy for fw3.

**Comment by bluewavenet** [comment]
@feckert 

> It is not recommended to mix iptables-nft and iptables-legacy commands.

But they both have exactly the same commands so you should be able to use either (but not both at the same time).

It is currently possible to install both (incorrectly) and it is very bad to use a mix of iptables-legacy and iptables-nft commands as you end up with two separate "iptables tables" with very unpredictable results...

So my question is: Can you run fw3 with iptables-nft?
I _think_ the answer is yes (once this PR is merged).

**Comment by champtar** [comment]
With iptables-legacy you had 1 table, doing ACCEPT was final
With nftables you have mutiple tables, doing ACCEPT continue the processing throught the remaining tables.
With fw3/iptables-legacy some software were inserting themselves early say in the INPUT chain, and doing an ACCEPT there, thus bypassing fw3 rules.
If you use fw3 + iptables-nft, previous example breaks.
if iptables-legacy doesn't provide iptables anymore, then people can't switch to fw3 + iptables-legacy without modifying the source and rebuilding themselves, that's a pretty high bar.

We just need to make sure ip(6)tables-nft is the default variant, haven't looked how to do it.
Also having both iptables-legacy and iptables-nft at the same time might allow to use iptables translate (not sure), in any cases I'm not a fan of conflicts.

**Comment by bluewavenet** [comment]
@champtar 

> I'm not a fan of conflicts.

Indeed, neither am I

> With iptables-legacy you had 1 table, doing ACCEPT was final
> With nftables you have mutiple tables, doing ACCEPT continue the processing throught the remaining tables.
> With fw3/iptables-legacy some software were inserting themselves early say in the INPUT chain, and doing an ACCEPT there, thus bypassing fw3 rules.

Yes. The openNDS package (and its predecessor NoDogSplash) did this and it was made more useful by adding the option to either ACCEPT or RETURN on each openNDS rule. 

>  If you use fw3 + iptables-nft, previous example breaks.

then you can choose to uninstall iptables-nft and install iptables-legacy if it is an issue in a particular edge case.

> We just need to make sure ip(6)tables-nft is the default variant, haven't looked how to do it.

This is the intended purpose of this PR...

**Comment by cotequeiroz** [comment]
If we want to keep the `PROVIDES` then I propose to get rid of the `IPTABLES_NFTABLES` option.  That dependency will not coexist the multiple `PROVIDES`.  The symbol is used to disable all nft packages with one click, but one can disable them individually, if desired.  It has no effect on the final binaries.  It will increase build time when you don't need the nft packages, but that usage shall be considered marginal anyway.

I still have to try this to be 100% certain, but I think we can rename the `*tables-nft` packages to just `*tables`, and have the `*tables-legacy` packages `PROVIDE` `*tables`.   Since [5936c4f libopkg: pkg_hash: prefer original packages to satisfy dependencies](https://git.openwrt.org/?p=project/opkg-lede.git;a=commitdiff;h=5936c4f9660248284e8a9b040ea3153d3ea888de), `opkg` installs the original package before searching for providers, so a dependency on just `iptables` will install the nft package, but do nothing if the legacy version is already installed.

**Comment by champtar** [comment]
@cotequeiroz looking at the commit you linked, renaming ip(6)tables-nft to ip(6)tables seems to be the best course of action
@feckert do you want to update this PR ?
- rename ip(6)tables-nft to ip(6)tables
- have ip(6)tables PROVIDES ip(6)tables-nft
- bump PKG_RELEASE

**Comment by cotequeiroz** [comment]
We should not forget about `arptables`, `ebtables`, and `nftables`.  Their alternatives have their own packages, but the issue is the same.

How about getting rid of `IPTABLES_NFTABLES`?

**Comment by bluewavenet** [comment]
@champtar @feckert @cotequeiroz 
Yes, make sure iptables-nft and ip6tables-nft are the defaults.
ie

- remove PROVIDES from legacy implementation as in original PR commit
- rename ip6tables-nft to ip6tables
- have ip6tables PROVIDES ip6tables-nft
- bump PKG_RELEASE

**Comment by champtar** [comment]
> @champtar @feckert @cotequeiroz Yes, make sure iptables-nft and ip6tables-nft are the defaults. ie
> 
>     * remove PROVIDES from legacy implementation as in original PR commit

No, already explained why in a previous message

**Comment by bluewavenet** [comment]
@champtar 

>     * remove PROVIDES from legacy implementation as in original PR commit
> No, already explained why in a previous message

If it is not removed then iptables-legacy will still be the default, as explained by @feckert in the first comment.
ie

> Since the firewall4 is now the default firewall in openwrt, the
> iptables-nft/ip6tables-nft backend for iptables must installed,
> if a package wants to use the iptables/ip6tables commands.

**Comment by champtar** [comment]
Have a look at @cotequeiroz message and the commit he linked

**Comment by cotequeiroz** [comment]
If you don't handle `IPTABLES_NFTABLES`, you end up with a ton of recursive dependencies, such as:
```
tmp/.config-package.in:1061:error: recursive dependency detected!
tmp/.config-package.in:1061:   	symbol PACKAGE_qos-scripts depends on PACKAGE_qos-scripts
```
At least it is easy to see their cause:
```
	config PACKAGE_qos-scripts
		select PACKAGE_iptables if PACKAGE_iptables-legacy<PACKAGE_qos-scripts
		depends on !(PACKAGE_iptables-legacy<PACKAGE_qos-scripts) || IPTABLES_NFTABLES
		
```
This happens because there's a conditional dependency on `IPTABLES_NFTABLES` if `iptables-nft` is to be selected.  The logic is that we should select iptables-nft if iptables-legacy is not selected.  Then if we are to select iptables-nft, we must add its dependencies to our list.  The selection condition includes the value of the symbol being defined, generating a recursive dependency.

It is a bug or limitation of `scripts/package-metadata.pl`.  The proper check for that condition should be done at the package selection, and there should be a line selecting the alternative should `IPTABLES_NFTABLES` not be `y`:
```
	select PACKAGE_iptables if (PACKAGE_iptables-legacy<PACKAGE_qos-scripts) && IPTABLES_NFTABLES
	select PACKAGE_iptables-legacy if !IPTABLES_NFTABLES
```
Notice that the selection of one package or the other depend on the value of `PACKAGE_qos-scripts`.

`IPV6` is thrown into a senseless depends line for a reason that I don't fully understand.
```
	depends on !(PACKAGE_iptables-legacy<PACKAGE_qos-scripts && IPV6) || IPV6
```
But here, the condition is always true, because if IPV6 is false, then the first part is always true; if not, the || condition will result in true as well.

**Comment by bluewavenet** [comment]
@champtar 
Do you mean:
> Since [5936c4f libopkg: pkg_hash: prefer original packages to satisfy dependencies](https://git.openwrt.org/?p=project/opkg-lede.git;a=commitdiff;h=5936c4f9660248284e8a9b040ea3153d3ea888de), opkg installs the original package before searching for providers, so a dependency on just iptables will install the nft package, but do nothing if the legacy version is already installed.

On a device flashed with current snapshot, as expected, iptables is not installed by default.
If I do `opkg install iptables` I get iptables-legacy.
So the above quote by @cotequeiroz  is unfortunately not what actually happens

**Comment by cotequeiroz** [comment]
Right now, there's no package named `iptables`.  It ends up in a "provider" list composed of `iptables-nft` and `iptables-legacy`.  opkg will pick one of them, which is not the one we want.  If we rename the package, then `iptables` will match an actual package, and the provider list won't be searched.

**Comment by champtar** [comment]
Because we haven't renamed iptables-nft to iptables yet ...

**Comment by bluewavenet** [comment]
@cotequeiroz 

> It is a bug or limitation of scripts/package-metadata.pl.

Now we are getting to the core of the problem.....

**Comment by bluewavenet** [comment]
> Because we haven't renamed iptables-nft to iptables yet ...

That would fix it nicely

**Comment by feckert** [comment]
@champtar @bluewavenet @cotequeiroz I have update the pullrequest with the suggested changes.

**Comment by champtar** [comment]
You forgot to bump PKG_RELEASE, else LGTM

**Comment by bluewavenet** [comment]
Looks good.
@cotequeiroz are you able to build and test?

**Review by cotequeiroz — CHANGES_REQUESTED** [review]
The `Package/*tables-nft/install` definitions need to be renamed as well.  Otherwise, LGTM.

**Comment by feckert** [comment]
@cotequeiroz Damn, I overlooked that -> done
@champtar bumped PKG_RELEASE as requested

**Review by cotequeiroz — APPROVED** [review]
Reviewed-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

**Comment by feckert** [comment]
@cotequeiroz would be nice if we get this into openwrt-22.03 before rc0

**Comment by cotequeiroz** [comment]
I don't have write access, so I can only agree. 😉

**Comment by cotequeiroz** [comment]
@aparcar, since you have conducted the migration to fw4, would you please take a look at this?

**Comment by jow-** [comment]
I don't think the renaming is required, if at all then it only adds confusion. The removal of  `IPTABLES_NFTABLES` makes sense. If the objective of the rename was to prefer `*-nft` oder `*-legacy` then use `DEFAULT_VARIANT:=1` on the `*-nft` package definitions.

**Comment by bluewavenet** [comment]
@jow- 
Whatever is the best way, we need the default to be iptables-nft and not iptables-legacy, along with preventing both being installed together.

**Comment by feckert** [comment]
> I don't think the renaming is required, if at all then it only adds confusion. The removal of `IPTABLES_NFTABLES` makes sense. If the objective of the rename was to prefer `*-nft` over `*-legacy` then use `DEFAULT_VARIANT:=1` on the `*-nft` package definitions.

@jow- I have now tried it out as you said. In my system, the mwan3 is the only one that has a dependency on [iptables](https://github.com/openwrt/packages/blob/master/net/mwan3/Makefile#L27) and [ip6tables](https://github.com/openwrt/packages/blob/master/net/mwan3/Makefile#L28).
If I add this change and regenerate my .config, then always iptables-legacy and ip6tables-legacy are enabled and not iptables-nft or ip6tables-nft!

Is it possible that the DEFAULT_VARIANT only works if I also define a build VARIANT?
https://github.com/openwrt/openwrt/blob/ec6293febc244d187e71a6e54f44920be679cde4/include/package-dumpinfo.mk#L27-L30

**Comment by pprindeville** [comment]
Blocking PR #18262 on merge...

**Comment by feckert** [comment]
@cotequeiroz @jow- frendly ping!
Any progress on this. It is still a problem vor me.

**Comment by cotequeiroz** [comment]
There's not much else I can do here.  I can run-test this again--when able--and report the results.

**Comment by cotequeiroz** [comment]
I've tested this in its current state.  If I install a dependent package, it pulls up `iptables-nft`.  However, if I try to install `iptables` directly, `iptables-legacy` is chosen: 😞

```
# opkg install sqm-scripts
Installing sqm-scripts (1.5.1-4) to root...
Downloading http://192.168.1.1/packages/aarch64_cortex-a53/packages/sqm-scripts_1.5.1-4_all.ipk
Installing iptables-nft (1.8.7-7) to root...
Downloading http://192.168.1.1/targets/mediatek/mt7622/packages/iptables-nft_1.8.7-7_aarch64_cortex-a53.ipk
Installing iptables-mod-ipopt (1.8.7-7) to root...
Downloading http://192.168.1.1/targets/mediatek/mt7622/packages/iptables-mod-ipopt_1.8.7-7_aarch64_cortex-a53.ipk
Configuring iptables-nft.
Configuring iptables-mod-ipopt.
Configuring sqm-scripts.
```

```
# opkg install iptables
Installing iptables-legacy (1.8.7-7) to root...
Downloading http://192.168.1.1/targets/mediatek/mt7622/packages/iptables-legacy_1.8.7-7_aarch64_cortex-a53.ipk
Configuring iptables-legacy.
```

**Comment by feckert** [comment]
@cotequeiroz The fact that iptable-nft is installed with an iptables [dependency](https://github.com/openwrt/packages/blob/68fb9c62a5f4421650aabb33c53466709afc46d6/net/sqm-scripts/Makefile#L27) is good so far.
That is what I want to achieve.
Unfortunately, I don't understand why it installs iptables-legacy when `opkg install iptables` is executed!

Actually, this should install iptables-nft because I defined `DEFAULT_VARIANT:=1` in my commit!
https://github.com/openwrt/openwrt/blob/0a48d333246a146445ee1a01157bca61814cca11/package/network/utils/iptables/Makefile#L138
https://github.com/openwrt/openwrt/blob/0a48d333246a146445ee1a01157bca61814cca11/package/network/utils/iptables/Makefile#L506

I need to check this again as I have only tried the dependency way.

Perhaps someone else from the core team can explain to us why this is so?
@adschm @nbd168

**Comment by hnyman** [comment]
> I don't understand why it installs iptables-legacy when opkg install iptables is executed!
>
> Actually, this should install iptables-nft because I defined DEFAULT_VARIANT:=1 in my commit!

I think that when the Makefile defines so many packages, it is unclear which packages the DEFAULT_VARIANT concerns. 

There is nothing to tie iptables-nft and iptables-legacy as being alternative variants, similarly regarding ip6tables-nft and ip6tables-legacy. I think that all packages defined in the Makefile are handled equally, and the last one with DEFAULT_VARIANT wins. That one is ip6tables-nft.

There is nothing to tie iptables-nft and iptables-legacy regarding being variants of each other, and neither is DEFAULT_VARIANT (as ip6tables-nft is). As neither is DEFAULT_VARIANT, the first one found providing is selected and can be iptasbles-legacy.

(Note: this is just a guess, I haven't yet checked the .packagedata in tmp/. )

**Comment by cotequeiroz** [comment]
Defining `DEFAULT_VARIANT:=1` will affect selection during `make menuconfig`, apparently.  The package information passed to `opkg` does not contain that information:
```
# opkg update
# zcat /tmp/opkg-lists/openwrt_core | grep -EA13 'Package: iptables-(nft|legacy)$'
Package: iptables-legacy
Version: 1.8.7-7
Depends: libc, xtables-legacy
Provides: iptables
Alternatives: 200:/usr/sbin/iptables:/usr/sbin/xtables-legacy-multi, 200:/usr/sbin/iptables-restore:/usr/sbin/xtables-legacy-multi, 200:/usr/sbin/iptables-save:/usr/sbin/xtables-legacy-multi
License: GPL-2.0
Section: net
CPE-ID: cpe:/a:netfilter_core_team:iptables
Architecture: aarch64_cortex-a53
Installed-Size: 243
Filename: iptables-legacy_1.8.7-7_aarch64_cortex-a53.ipk
Size: 1206
SHA256sum: 357584a46978699ab2017cb60719a076227d386a88e48e848361377be1d1a148
Description:  IP firewall administration tool.
--
Package: iptables-nft
Version: 1.8.7-7
Depends: libc, kmod-ipt-core, xtables-nft
Provides: iptables
Alternatives: 300:/usr/sbin/iptables:/usr/sbin/xtables-nft-multi, 300:/usr/sbin/iptables-restore:/usr/sbin/xtables-nft-multi, 300:/usr/sbin/iptables-save:/usr/sbin/xtables-nft-multi
License: GPL-2.0
Section: net
CPE-ID: cpe:/a:netfilter_core_team:iptables
Architecture: aarch64_cortex-a53
Installed-Size: 244
Filename: iptables-nft_1.8.7-7_aarch64_cortex-a53.ipk
Size: 1065
SHA256sum: 88500e71537d6fbee41cba40e3a6303c8574dad900689f802a129b71234784f4
Description:  Extra iptables nftables nft binaries.
```
I don't have the time to look into this much deeper.  Apparently, `opkg` has to decide which package to install based on the contents of files under `/tmp/opkg-lists` (it does not download every package providing `iptables` to peek at their metadata).  Unless we change the way this works--at least making the behavior of installing dependent packages match the command line action in a predictable way--we may need to rely on the name preference I mentioned above.

**Comment by cotequeiroz** [comment]
> I think that when the Makefile defines so many packages, it is unclear which packages the DEFAULT_VARIANT concerns.

`DEFAULT_VARIANT` does make its way to `$(TOPDIR)/tmp/.packagedata`:
```
$ sed -n '/Package: iptables-\(nft\|legacy\)$/,/Description:/p' tmp/.packageinfo
Package: iptables-legacy
Submenu: Firewall
Version: 1.8.7-7
Depends: +libc +USE_GLIBC:librt +USE_GLIBC:libpthread +xtables-legacy
Conflicts:
Menu-Depends:
Provides: iptables
Section: net
Category: Network
Title: IP firewall administration tool
Maintainer:
Source: iptables-1.8.7.tar.bz2
License: GPL-2.0
Type: ipkg
Description: IP firewall administration tool.
Package: iptables-nft
Submenu: Firewall
Version: 1.8.7-7
Depends: +libc +USE_GLIBC:librt +USE_GLIBC:libpthread +kmod-ipt-core +xtables-nft
Conflicts:
Menu-Depends:
Provides: iptables
Build-Variant: nft
Default-Variant: nft
Section: net
Category: Network
Title: IP firewall administration tool nft
Maintainer:
Source: iptables-1.8.7.tar.bz2
License: GPL-2.0
Type: ipkg
Description: Extra iptables nftables nft binaries.
```
If I add a variant to the legacy version, it will just add a `Build-Variant:` line below `Provides`, and does not change the `opkg` metadata.

**Comment by champtar** [comment]
Another proposal to fix this issue #10124

**Comment by feckert** [comment]
Superseeded by https://github.com/openwrt/openwrt/pull/10124

**Pull request closed without merging**
